File: Python3.x/206-Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        curr, prev = head, None
        while curr:
            next = curr.next  # 下一节点就是现在的下一个节点
            curr.next = prev  # 现在的下一个节点指针指向前一个节点
            prev = curr  # 前一个指针后移
            curr = next  # 当前指针后移
        return prev  # 最前面节点


'''
三个指针：当前节点、后面节点、前面节点
'''

File: C++/206-Reverse_Linked_List.cpp
// Runtime: 8 ms, faster than 85.27% of C++ online submissions for Reverse Linked List.
// Memory Usage: 9.1 MB, less than 73.82% of C++ online submissions for Reverse Linked List.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *cur = NULL;
        while (head) {
            ListNode *next = head -> next;
            head -> next = cur;
            cur = head;
            head = next;
        }
        return cur;
    }
};
